### PR TITLE
Add GPT2 model

### PR DESCRIPTION
**Pull request by janEbert**
I ported this from another project for which I used an older Flux version (still with `@treelike` and without Zygote). Sadly, I am not able to run the test suite as I am getting CUDA segfaults even without the tests I added (I listed the stacktrace below; this is probably an issue on my part). However, manually executing the test file reveals the same error the old `Gpt` model also has (the very last test fails).

I still have to document the methods but maybe you can already give me some feedback. I have _not_ added support for pre-trained models but maybe I can do that in the future as well. No promises, though. :)

```julia
(Transformers) pkg> test Transformers
   Testing Transformers
 Resolving package versions...
[ Info: Test CUDA
[ Info: Testing CUDA

signal (11): Segmentation fault
in expression starting at /home/jan/.julia/dev/Transformers/test/cuda/gather.jl:1
_ZN4llvm15ValueHandleBase12AddToUseListEv at /home/jan/Downloads/julia-1.3.1/bin/../lib/julia/l
ibLLVM-6.0.so (unknown line)
_ZNSt6vectorIN4llvm14WeakTrackingVHESaIS1_EE12emplace_backIJRPNS0_5ValueEEEEvDpOT_ at /home/jan
/Downloads/julia-1.3.1/bin/../lib/julia/libLLVM-6.0.so (unknown line)
double free or corruption (out)

signal (6): Aborted
in expression starting at /home/jan/.julia/dev/Transformers/test/cuda/gather.jl:1
```

### CODE REVIEW

File: src/gpt2/model.jl [hunk 1]
```diff
@@ -0,0 +1,194 @@
...
+                              return_all_outputs::Bool=false)
+    input = input .+ gpt.embedding(input)
+    output = gpt.gpt(input, mask, return_all_outputs=return_all_outputs)
+    return gpt.output(output)
+end
+
+
+# Layer initialization
+
+init_weights!(layer::GPT2Predictor, std=0.02)       = init_field_weights!(layer, std)
```
Comment (chengchingwen) [inline src/gpt2/model.jl]: Why do we need these init functions?

**Comment by chengchingwen** [comment]
Hi! 
First of all, thanks for the contribution. It's glad to see something/someone new coming in. 

However, I'm a little busy these days, so I won't be able to do anything until Feb. 9. There are still some bugs with migrating to Zygote (That's why you see errors with the GPT test). I'll start fixing the problems and probably do some refactoring if needed when I'm free. 

Last but not the least, the coding style are not consistent with the codebase. You probably want to take some time on that.

File: src/gpt2/model.jl (hunk 1 above)
Comment (janEbert) [reply src/gpt2/model.jl]: The [original](https://github.com/openai/gpt-2/blob/0574c5708b094bfa0b0f6dfe3fd284d9a045acd9/src/model.py#L50-L56) source [code](https://github.com/openai/gpt-2/blob/0574c5708b094bfa0b0f6dfe3fd284d9a045acd9/src/model.py#L28-L38) as well as [HuggingFace](https://github.com/huggingface/transformers/blob/ea2600bd5f1d36f2fb61958be21db5b901e33884/src/transformers/configuration_gpt2.py#L131) use initialization functions with (mostly) the given default values.

Now that I looked over the code again, perhaps the `PositionEmbedding` [should have](https://github.com/openai/gpt-2/blob/0574c5708b094bfa0b0f6dfe3fd284d9a045acd9/src/model.py#L152-L153) 0.01 as initalization stderr.

Once again, sorry about no documentation yet; I'll fix it as soon as possible.

**Comment by janEbert** [comment]
Hey Peter, thank you as well for such a great library!
I'm also busy right now so sorry if improvements take me some time. I'll be sure to do it as soon as I can. And good luck with whatever keeps you busy at the moment!
I understand having problems with Zygote bugs all too well; hope it's not too much work.

Can you please tell me what about the coding style is inconsistent? Some things I quickly saw is
1. the usage of `import` instead of `using`,
2. naming it `GPT2` instead of `Gpt2` (unlike the old `Gpt` model),
3. differently named variables,
4. slightly different type signatures (source: `Int` -> PR: `Integer`) and
5. using `Float32`s by default.
Anything else? I'd argue 2., 3. and 4. were objectively good changes and if you want, I can adapt the rest of the codebase. However, I also understand the delicacy of this topic so just tell me if you don't want any of those changes.

Once again, I'm sorry the PR is such a messy submission; I should have done a better job before submitting but as I said, I really wanted some feedback before putting too much time into it. I'm also glad that you are happy about the contribution!

File: src/gpt2/model.jl [hunk 2]
```diff
@@ -0,0 +1,194 @@
...
+        res_attn = reshape(res_attn, insize[1], :)
+    end
+    res_attn = b.attn_norm(res_attn)::typeof(res_attn)
+    pwff = b.piecewise(res_attn)
+    pwff = b.dropout(pwff)
+    res_pwff = res_attn .+ pwff
+    if N == 3
+        res_pwff = reshape(res_pwff, :, Base.tail(insize)...)
+    end
+    res_pwff
```
Comment (chengchingwen) [inline src/gpt2/model.jl]: Could you check this implementation with the origin(openai) again? I only scan through, but it doesn't look correct.

File: src/gpt2/model.jl (hunk 1 above)
Comment (chengchingwen) [reply src/gpt2/model.jl]: I mean, these functions shouldn't be in this module or even this package. They are the util functions of Flux. If we don't have any special usage that require these, then we probably don't want them.

**Comment by chengchingwen** [comment]
Don't worry about the time and docs, do it with your own pace.

Here are some stuffs I think that it can improve.

1. some function/module should be imported into this module. Probably check how and what `gpt/GenerativePretrain.jl` import/export.
2. naming: `Gpt2` as the struct, `GenerativePretrain2` as the module.
3. using `Float32` as default is very import since that's the default of Flux.
4. We don't really need the `GPT2Predictor` because currently I use the `TransformerModel` struct for the similar usage (see the examples). It might not be a good api, but for now let's follow the old code. (If you have something better in mind, just speak out. we can move the codebase to the better one in another PR.)

5. IIRC, the `GPT2Block` is the Post-LN transformer in [this paper](https://openreview.net/pdf?id=B1x8anVFPr). If it is the case, it's better to move the code to `basic/transformer.jl` and rename as something like `PostLNTransfomer`

File: src/gpt2/model.jl (hunk 2 above)
Comment (chengchingwen) [reply src/gpt2/model.jl]: see here
https://github.com/openai/gpt-2/blob/0574c5708b094bfa0b0f6dfe3fd284d9a045acd9/src/model.py#L123-L130

**Comment by chengchingwen** [comment]
I was planning about adding these stuffs myself after I fix the Zygote bugs, but I got too busy to do that. Here is the only file I wrote that time. You can take it as an example.
```jl
# in gpt2/GenerativePreTrain2.jl
module GenerativePreTrain2


using Flux
using Requires
using Requires: @init
using BSON

using ..Transformers: Abstract3DTensor
using ..Basic
using ..Pretrain: isbson, iszip, isnpbson, zipname, zipfile, findfile

export Gpt2, load_gpt2_pretrain

include("./gpt2.jl")
include("./load_pretrain.jl")

end
```

**Comment by janEbert** [comment]
Hey, I'm sorry for the long delay. I'll try to fix everything until Monday.

**Comment by janEbert** [comment]
This will take longer; I leaned way too much on the HuggingFace implementation. I see now what you meant about the calling function for `Gpt2Block`. The OpenAI implementation does not even have dropout anywhere (even though it is mentioned in the original GPT paper).
I'll try to get this merge-ready this week instead.

**Comment by janEbert** [comment]
Please take another look. I tried to incorporate all your feedback and rebased on master.
Sadly, I cannot load the project due to an error with CuArrays.jl:
```julia
julia> using Transformers
[ Info: Precompiling Transformers [21ca0261-441d-5938-ace7-c90938fde4d4]
ERROR: LoadError: LoadError: syntax: invalid keyword argument syntax "(escape (= threads nthreads))"
[...]  # four stacktraces
```
I think some docstring references will not work currently but I can't test it.

**Comment by chengchingwen** [comment]
looks like something crash with CuArrays, but I can't reproduce the error. What's your version?

**Comment by janEbert** [comment]
Julia 1.4.0, CuArrays 1.6.0

**Comment by chengchingwen** [comment]
I’m using Julia 1.4 with CuArrays 1.7.3. Maybe try to update CuArrays to 1.7.3 (btw, Transformers is now 0.1.3)

**Comment by janEbert** [comment]
Actually, Transformers itself has a compat bound on CuArrays 1.6 in Project.toml. :)
But I'll try to upgrade independently.

**Comment by chengchingwen** [comment]
The compat bound on CuArrays 1.6 because I was using and testing with Julia 1.3

**Pull request closed without merging**